### DISCUSSION TITLE
New Releases OC 4 & 3 ??

### DISCUSSION THREAD

**Discussion by osworx** [opencart/opencart#13993] [General]
What prevents Daniel to publish new releases of 3.0.4.0 and 4.1.0.0 ???

3.0.4.0 is ready for production since weeks!
And 4.1.0.0 would be fine to have a dev-version which will be freezed (2-4 weeks) until last bugs are removed.
Adding always new functions, changing routines in 4.x is useless after that time.

**Comment by chaoos2** [comment]
I am interested in version 3.x, I will be happy with the new release

**Comment by felipoantonoff** [comment]
The important thing for Opencart 4 is to remain as stable as possible to be more adopted and the return of OCMOD to begin to see good improvements and become more popular.
For now, Opencart 3 is still the main current Opencart and it has been more than 1 year since Opencart 4 was launched.

Features can be thought of after the fixes and some can be left for modules/apps.

Opencart 3 I found it strange to continue, but I understand that this is due to Opencart 4 not being so stable yet and many who are on 3 needing support for PHP 8.

**Comment by mhcwebdesign** [comment]
OCMOD is not vital, unlike what you think. And PHP 8 versions are well supported by OpenCart 3, e.g. 3.0.3.9 or 3.0.4.0.

**Comment by mhcwebdesign** [comment]
OpenCart 3.0.4.0 has now been tagged and released by Daniel, see https://github.com/opencart/opencart/tree/3.0.4.0

**Comment by andresjavierpf** [comment]
Is there any information on when the next version of OC 4 will be released? 
Is the current version 4.0.2.3 stable for use in production, and have the translation issues been resolved? 

Does anyone know anything?

**Comment by felipoantonoff** [comment]
@andresjavierpf 4.1 was recently released, but it still lacks some improvements related to OCMOD and some other details.

**Comment by mhcwebdesign** [comment]
While there have been some improvements with regards to OCMOD in the current master branch (4.1.0.1), some new changes were introduced, and chances are there will be more bugs. I just installed the current master branch on March 20, 2025, and it took me only a few minutes of quick testing to come across 4 different bugs (https://github.com/opencart/opencart/issues/14698, https://github.com/opencart/opencart/issues/14697, https://github.com/opencart/opencart/issues/14696, https://github.com/opencart/opencart/issues/14695). 

That's why I think that the next release should be a public beta-release, it should not be declared as a stable release yet. And during the public beta release, no new features should be introduced (as happened again with 4.1.0.1 vs 4.1.0.0), but the focus should sole be on bugfixes!

**Reply by felipoantonoff** [reply]
I agree, it needs to be better tested, especially the fundamental features such as Registration, Login and of course the ability to generate an Order.

I installed 4.1 recently and had problems even doing a simple Registration. We need to learn from Opencart 4, which was a very early release and without proper testing.

I even suggest automated tests simulating real use, with Playwright, Puppeteer or something similar, as this would easily catch several bugs and make future testing easier.

Opencart 4 should now focus on becoming stable and not on features. Many features should be left to plugins/apps, otherwise the core/base could become too bloated and also more complex to maintain, something that happened with some Opencart competitors.

Now it would be good for Opencart to have the complete OCMOD, so that those who like it can use it and those who prefer another option can use Events or use other plugins and it needs to be very stable, until today, years after Opencart 4, I have only been recommending Opencart 3 and I hope one day I can recommend Opencart 4 as the main option instead of 3, which I consider much more stable and mature.

**Comment by BlackSoll21** [comment]
Yes, we all need a rock solid bug free 4.1.x.x version for production. Any new features are welcomed but who cares if version is not stable for a live site ?

**Reply by felipoantonoff** [reply]
Exactly, the focus now has to be on a stable Opencart 4 ready for production, too many resources can also make it more difficult to maintain and can bring more possible bugs, as much as many people like new things, I prefer a good strong and simple base as Opencart has always been and new resources in general through Plugins/apps, so each Store uses what it really needs without bloating the system, as happened with several Opencart competitors.

**Comment by pemaier** [comment]
A new release (OpenCart 4.1.0.1 is available.
And with the now starting development process we solve 2 goals.
1. Stability within the master branch
2. New feature developments in the development branch

I guess @danielkerr is starting asap a process to establish a regular update process from the master branch to the development branch. (That's required to transport the features into the master branch in a working condition.)

**Comment by danielkerr** [comment]
yes i am. but i also need the help fixing stuff. its stopping me developing new stuff. in w9lling to pay. i need one full time fixer.

**Reply by felipoantonoff** [reply]
I understand that there is a lot to do, but I would recommend pausing new features and focusing on stability and having the complete OCMOD as in Opencart 3 so that developers can start updating their Plugins for it, at least until Opencart 4 becomes stable enough for production and development.

Opencart is already feature-rich and Opencart 4 has added many new features, and many of them can be installed by Apps so that the base does not become too bloated and difficult to maintain.

I would recommend automating the tests. As soon as I have more time, I will try to do some tests with Playwright, Puppeteer or some system like that. They can simulate the use of the site, such as testing Registration, Login, Cart, Purchase of an Order, etc. This way, once implemented, it would be easier to do the main tests that today would be done more manually, and this way it would be possible to discover more bugs and fix them. It would also be possible to take Screenshots.

**Reply by stalker780** [reply]
I also recommend to take a brake for 2-3 months to make a stable OC4.
In the last 3 years there were no stable OC4 releases. 
All releases contained critical bugs and each release introduced new features which break any extensions compatibility against previous version.

OC4 is good enough and releases should pay more attention to stability. Extensions developers cannot concentrate and anchor to one version now.

Perhaps OC github should snapshot OC4 and DB structure as a stable version which will get fixes and libraries updates next several years.
And start a new OC5 as a beta master branch with headless and other features.

Progress and ideas never stops. But we need to live right now. Lets just move step by step. OC4 is a step. Just fix it.

**Reply by stalker780** [reply]
@danielkerr  In my opinion, each OC version should have determined lifecycle. Just take the Linux idea, for example CentOS/Alma (10 years) vs Fedora (1 year).

1. Major/master stable version lives 5-10 years and gets security and library updates till it is possible against PHP and MySQL versions compatibility. This is an LTE version like OC 2.3 or 3.0 were. They are still used right now even on PHP 5.6. Cause there is nothing stable to replace them.

2. Beta version is developed 1-3 years. For enthusiasts of edge technologies. After years of tests and experiments it becomes a major version.

3. Alpha version for experiments. For geeks. Many features will never see daylight, but still some ideas may go to beta and then to master.

Please do not transform OpenCart into a waterfall of neverending new features and bugs which lasts for decades.

PS. OC3 still got some fixes and updates to PHP8... But look when BS3 was released. It was more than 10 year ago. It is outdated more than completely. And we have to support it only because there is no stable inheritor in these 10 years.

**Comment by pemaier** [comment]
I do not see the requirement to stop development new features for a time.
We are all developers with different competences and the knowledge to work with rules.
So all know the best practices for development processes and use them already in small teams and/or big companies in daily usage.

We do not need to develop someting new. Let's integrate our experiences as Anton @stalker780 already mentioned.
To start small but with a big effect I see in the following steps. It should be possible to integrate them within the next 2-3 weeks.

**Decoupling of development and maintenance code lines and establish a transport system for a ongoing development process.**
1. Setup a development branch (development already exist)
This branch us used for development of new features.

2. Establish a language support branch (I10n already exist)
This branch us used for language themes as @osworx is involved in with Crowedin and for automatic builds and download area of language extensions.
(more details are already in discussion with @osworx and @danielkerr I guess)

3. Establish an integration branch
This branch us used for integration of several development branches. Right now there is development branch and I10n branch (language). 
May we also have to establish in the future more branches. 

4. Establish a testing branch
This branch us used for testing before the new developments are getting transported into and the implementation for emergency fixes. 
It is the area where the test teams are having test periods before new releases.


**Build up automatic software transport lines.**

**Code transport lines** (branches)
**Development  -->  Integration  -->  Testing  -->   Master**

![image](https://github.com/user-attachments/assets/72559384-fa0d-4789-a25d-7bde8b44ac5b)




**It's just a proposal.**

**Comment by rsuchovsky** [comment]
Is it possible to give some information on when the new version oc 4.1.0.4 will be released? At least whether the estimated time is within a week or a month...

**Comment by danielkerr** [comment]
its not to far off. i'm just finishing up a new feature then i need to go through the issue list again. i can normally get through a lot of issues in a week.

**Reply by osworx** [reply]
> its not to far off. i'm just finishing up a new feature then i need to go through the issue list again...

That's exactly the problem!

Releasing a new version should be done asap.
Then - and only then !! - based on that **stable release** adding new features can be added as many as you want.

As I suggested many years ago (and now others are doing), establish a **Release Managerment** if **no** roadmap may be available.

Second, create something like a **freeze periode** -> means a timeframe (2-6 weeks) the branch is **freezed** and the developers and the **users can test and eleminate last bugs**.

Third, after this ** freeze periode** the release if the **final and stable release** can be made.

Fourth: adding new features can be made in a **development or feature branch** - at any time.
Is the teature (or are the features) ready to test, the same loop starts again.

With such a structure everybody knows what's going on.
Can count on that, can create and adopt own extensions and has a positive view on that whole thing ..

nb: what I am missing (maybe other also) is, what **new feature(s)** you are adding and what do you plan in future?
For example **SSR** maybe good, but what's the reason and benefit of it (without looking into the code) ?

**Comment by mhcwebdesign** [comment]
Could you make a public beta release first, for a few weeks, before a final 4.1.0.4 ?

**Comment by danielkerr** [comment]
we dont have time for this sort of thing. let me get this next feature finish then i will be back to doing quick releases with just a few fixes in.

**Reply by osworx** [reply]
Sorry Daniel, but don't say we "have no time for this .." 

We - the devs and user - have time to test new relaseses **before" they will be published!
Because the forum and the issue tracker here is always full of reports "what's not working, not functional, etc .." after a new release.

If you have the time to add new features, we have the time to test > simple case.
And it makes absolutely no sense to publish a fewnew releases within a few days as we had that already in the past because the latest version was full of bugs!

**Comment by felipoantonoff** [comment]
I think new features are interesting, but what makes Opencart the best e-commerce system is its lightweight core/base, easy to use, easy to customize, low resource consumption and excellent usability in general for the end user and administrator.

We have to be careful not to focus too much on new features and turn Opencart into another e-commerce platform. Opencart has never focused on being bloated or having thousands of features. Those who want extra features use Apps/Plugins or create custom ones with their developer. My biggest client has been doing this for years. We create new features according to their needs, which are very specific.

The resources of an e-commerce must be the essentials, so as not to make it bloated, complex and heavy for the server and database. Opencart already has the essentials and much more natively, the rest is improved with Plugins and Apps from good professionals. After all, it is impossible to have every type of resource natively, which would make maintenance unsustainable, and too many resources increase the likelihood of bugs and more work to update to future versions.

In my view, Opencart must remain an e-commerce focused on its core/base and leave it to the Developers to create extra Plugins and Apps. This keeps the ecosystem around it strong and avoids having a version with many bugs or heavy weight, after all, there are many resources that are not used or needed by everyone.

An example is native Payments and Deliveries, which the vast majority of people do not even use depending on the country. For example, in Brazil, they will use Apps made by other companies and professionals, due to the particularities of the Brazilian market, payments and deliveries. I understand that we can choose not to activate it, but believe me, there are customers who would like to clean it up and leave only what they really use. Of course, this is a separate case, but it shows how many features it has.

Opencart 4 really focused on having a new Checkout, which I think is interesting, but ugly with many bugs. Wouldn't it be better for those who want a One Checkout to use an already established and well-tested App for this purpose, and those who want a simple Checkout in stages should use the native one. Of course, it is a totally valid feature, but any new feature should come after Opencart 4 is very stable. The focus should always be on stability and maturity, in order to receive more support from developers and companies, since most continue to support Opencart 3, which is still the best current version of Opencart and the easiest to work with, and the one with the most stable OCMOD, since the current Opencart 4 has limited OCMOD, one of the biggest differences in Opencart 4.

I sometimes get asked which Opencart to use, and unfortunately I have to recommend Opencart 3. I suggest that Opencart 4 should become more mature, more stable, and with more good apps within about 6 months.

It would be interesting to have a well-updated Roadmap and perhaps first focus on making it stable and then on new features or perhaps in parallel, but with the priority always being stability, after all, there is no point in an E-commerce full of features if the customer cannot buy or have a good shopping experience and often less is more, even more so in E-commerce and this being said by a Developer who has already made more than 100 Apps and Systems for Opencart and even so I recommend using each feature according to the need.

**Comment by cgcmart** [comment]
@FelipoAntonoff 

> first focus on making it stable and then on new features or perhaps in parallel, but with the priority always being stability, after all, there is no point in an E-commerce full of features if the customer cannot buy or have a good shopping experience 

I totally agree. Briefly, an OC 4 stable version for production is needed and should be focused.

**Comment by danielkerr** [comment]
maybe break the versions into development and master branch. i leave the master branch alone and let people post fixes there.

**Reply by mhcwebdesign** [reply]
I'd be happy to post fixes into it. But at least could you make sure the installer is actually fixed first? 

The **system/library/db.php**, along with **system/library/db/mysqli.php** and **install/model/install.php** are all broken at the moment!

**Reply by pemaier** [reply]
When a working install (incl. install_cli) is expected?

**Comment by rsuchovsky** [comment]
Is it possible to give a release date for the new version 4.1.0.4?
I also want to thank you for your work!

**Comment by danielkerr** [comment]
i think maybe another 2 weeks.

**Comment by rsuchovsky** [comment]
@danielkerr , can you give a realistic date for the release of the new version of opencart 4?
I know that there are a lot of adjustments and changes being made, that's why the date has been pushed back.
I appreciate your work. I would like to know if you have any idea about a possible release. TXH

**Comment by totalsport-tech** [comment]
I also think it’s too much we need to fix.
https://totalsport.ua/futbol-c-31/inventar-c-31_328/

**Comment by felipoantonoff** [comment]
Hi everyone, I created a thread about the future of Opencart and suggestions, anyone who wants to participate:
https://github.com/opencart/opencart/discussions/15300

@danielkerr @osworx
